src/simulators/tensor_network/MPS.cpp

```diff
@@ -0,0 +1,398 @@
+	right_gamma.div_Gamma_by_right_Lambda(right_lambda);
+	q_reg_[index_A] = left_gamma;
+	lambda_reg_[index_A] = lambda;
+	q_reg_[index_B] = right_gamma;
+}
+
+void MPS::change_position(uint_t src, uint_t dst)
+{
+	if(src == dst)
+		return;
+	else if(src < dst)
+		for(uint_t i = src; i < dst; i++)
+			apply_swap(i,i+1);
+	else
+		for(uint_t i = src; i > dst; i--)
+			apply_swap(i,i-1);
+}
+
+cmatrix_t MPS::Density_matrix(const reg_t &qubits) const
+{
+  // ***** Assuming ascending sorted qubits register *****
+  vector<uint_t> internalIndexes;
+  for (uint_t index : qubits)
+    internalIndexes.push_back((uint_t)index);
+  //  std::sort(internalIndexes.begin(), internalIndexes.end()); -- Assuming sorted
+
+  MPS temp_TN;
+  temp_TN.initialize(*this);
+  vector<uint_t> new_indexes = calc_new_indexes(internalIndexes);
+  uint_t avg = new_indexes[new_indexes.size()/2];
+  vector<uint_t>::iterator it = lower_bound(internalIndexes.begin(), internalIndexes.end(), avg);
+  int mid = std::distance(internalIndexes.begin(), it);
+  for(uint_t i = mid; i < internalIndexes.size(); i++)
+  {
+    temp_TN.change_position(internalIndexes[i],new_indexes[i]);
+  }
+  for(int i = mid-1; i >= 0; i--)
+  {
+    temp_TN.change_position(internalIndexes[i],new_indexes[i]);
+  }
+  MPS_Tensor psi = temp_TN.state_vec(new_indexes.front(), new_indexes.back());
+  uint_t size = psi.get_dim();
+  cmatrix_t rho(size,size);
+  for(uint_t i = 0; i < size; i++) {
+    for(uint_t j = 0; j < size; j++) {
+      rho(i,j) = AER::Utils::sum( AER::Utils::elementwise_multiplication(psi.get_data(i), AER::Utils::conj(psi.get_data(j))) );
+    }
+  }
+  return rho;
+}
+
+double MPS::Expectation_value(const reg_t &qubits, const string &matrices) const
+{
+  // ***** Assuming ascending sorted qubits register *****
+  cmatrix_t rho = Density_matrix(qubits);
+  string matrices_reverse = matrices;
+  reverse(matrices_reverse.begin(), matrices_reverse.end());
+  cmatrix_t M(1), temp;
+  M(0,0) = complex_t(1);
+  for(const char& gate : matrices_reverse)
+  {
+    if (gate == 'X')
+	  temp = AER::Utils::Matrix::X;
+    else if (gate == 'Y')
+	  temp = AER::Utils::Matrix::Y;
+    else if (gate == 'Z')
+	  temp = AER::Utils::Matrix::Z;
+    else if (gate == 'I')
+	  temp = AER::Utils::Matrix::I;
+    M = AER::Utils::tensor_product(M, temp);
+  }
+  // Trace(rho*M). not using methods for efficiency
+  complex_t res = 0;
+  for (uint_t i = 0; i < M.GetRows(); i++)
+    for (uint_t j = 0; j < M.GetRows(); j++)
+      res += M(i,j)*rho(j,i);
+  return real(res);
+}
+
+double MPS::Expectation_value(const reg_t &qubits, const cmatrix_t &M) const
+{
+  // ***** Assuming ascending sorted qubits register *****
+  cmatrix_t rho = Density_matrix(qubits);
+
+  // Trace(rho*M). not using methods for efficiency
+  complex_t res = 0;
+  for (uint_t i = 0; i < M.GetRows(); i++)
+    for (uint_t j = 0; j < M.GetRows(); j++)
+      res += M(i,j)*rho(j,i);
+  return real(res);
+}
+
+void MPS::printTN()
```

Comment:
As I have commented above, overloading the << operator is considered a good practice for gathering debugging information from a class. It will return a `std::ostream` that you can use later within a `std::cout` for example.

Reply:
done.